tlv/varint.go

```diff
@@ -0,0 +1,109 @@
+package tlv
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+)
+
+// ErrCanonicalVarInt signals that the decoded varint was not minimally encoded.
+var ErrCanonicalVarInt = errors.New("decoded varint is not canonical")
+
+// ReadVarInt reads a variable length integer from r and returns it as a uint64.
+func ReadVarInt(r io.Reader, buf *[8]byte) (uint64, error) {
```

Comment:
yes, and quite a significant one. passing `buf` into an `io.Writer` will cause it to escape to the heap. each [de]-serialization of a varint would then allocate 8-bytes to the heap , rather than one for the entire stream that can be reused for encoding/decoding.